src/io/mod.rs

```diff
@@ -1,5 +1,123 @@
 //! Input and output of images.
+
+use crate::{error, ImageError, ImageResult};
+
 mod reader;
 pub(crate) mod free_functions;
 
 pub use self::reader::Reader;
+
+/// Set of supported strict limits for a decoder
+#[derive(Clone, Debug, Eq, PartialEq, Hash)]
+#[allow(missing_copy_implementations)]
+pub struct LimitSupport {
+    _non_exhaustive: (),
+}
+
+impl Default for LimitSupport {
+    fn default() -> LimitSupport {
+        LimitSupport {
+            _non_exhaustive: (),
+        }
+    }
+}
+
+/// Resource limits for decoding.
+///
+/// Limits can be either *strict* or *non-strict*. Non-strict limits are best-effort
+/// limits where the library does not guarantee that limit wont be exceeded. Do note 
+/// that it is still considered a bug if a non-strict limit is exceeded, however as 
+/// some of the underlying decoders do not support not support such limits one cannot 
+/// rely on these limits being supported. For stric limits the library makes a stronger 
+/// guarantee that the limit will not be exceeded. Exceeding a strict limit is considered 
+/// a critical bug. If a decoder cannot guaranree that it will uphold a strict limit it 
+/// *must* fail with `image::error::LimitErrorKind::Unsupported`.
```

Comment:
Any suggestions for alternate wording here. If we go ahead and merge this PR without any strict limits this may be slightly confusing?

Reply:
Maybe this should allude to the potential to introduce strict limits in the future, noting that the `LimitSupport` will default to false, and that decoders should turn them to true if they can consider their violation a critical bug.